Add _apply_list_op, each, map, filter to ComplexBase

cifkao opened the pull request:
- Adding an `_apply_list_op` method that recursively applies a list operation to list attributes.
- Implementing `remove_duplicate`, `remove_invalid` and `sort` using this method.
- Adding `each`, `map` and `filter` functions.

Examples:
```python
def is_in_range(obj, a, b):
    if hasattr(obj, "time"):
        return obj.time >= a and obj.time <= b
    return True

# Crop to a given range
music.filter(functools.partial(is_in_range, a=4, b=8)).adjust_time(lambda t: t - 4)
```

```python
def fix_pitches(obj):
    if hasattr(obj, "pitch") and obj.pitch < 0:
        obj.pitch += (obj.pitch + 11) // 12 * 12

# Transpose and fix invalid pitches
music.transpose(-60).each(fix_pitches)
```

cifkao commented:
Fixes #49 (seemingly).

cifkao commented:
Related to #32.

salu133445 commented:
The functions `map`, `filter` and `each` can be quite handy, although they look pretty non-pythonic to me.

For `map` and `each` function, I think a more pythonic way is to have a for loop iterating over all the objects regardless of their types (similar to `music21.score.flat`). We could have a method `Music.objects()` that returns an iterator on all the objects. Then, we can do something as follows, which seems more pythonic to me.

```python
for obj in music.objects():
    fix_pitches(obj)
```

(We cannot achieve the `filter` function as above though as it needs to modify the lists.)

cifkao commented:
`map` is still useful, because it replaces values with new ones (maybe you can think of a name for it that better reflects this functionality). But replacing `each` with a flat iterator sounds nice. `objects` sounds quite generic, I think `flat`, like in numpy or music21, is better.

salu133445 commented:
Oops. You are right. `map` does not work that way. We could have a list iterator and do as follows.

```python
for list_ in music.lists():
    list_[:] = map(func, list_)
```

Similarly, for `filter`, we could do the following.

```python
for list_ in music.lists():
    list_[:] = [obj for obj in list_ if is_in_range(obj, a=4, b=8)]
```

Neither is that pythonic though.

salu133445 commented:
Also, isn't `map` more like `apply` in PyTorch? Python's built-in `map` returns a new iterable, but the one in this pull request does modify the lists in place.

salu133445 commented:
`Music.objects()` is probably not a good one as it sounds like everything, including non-list attributes, will be included. `Music.flat()` sounds good to me.

salu133445 commented:
It seems that `Music._apply_list_op()` is rather powerful. Why don't we make it `Music.apply()`? This could perhaps simplify `map` and `filter` by using `apply` with built-in `map` and `filter`.

cifkao commented:
> Also, isn't `map` more like `apply` in PyTorch? Python's built-in `map` returns a new iterable, but the one in this pull request does modify the lists in place.

Yes, a bit, but `apply` in PyTorch also doesn't replace the values. This `map` is like `map` but in-place, I'm not sure what's a good name for it. Maybe `apply` would work actually?

> It seems that `Music._apply_list_op()` is rather powerful. Why don't we make it `Music.apply()`? This could perhaps simplify `map` and `filter` by using `apply` with built-in `map` and `filter`.

I was thinking about that. But the thing is that it accepts a function with 3 different parameters (of which I'm currently using only 2 actually...) and so (1) it can't be used with existing functions like built-in `map` and `filter` and (2) I thought it would be difficult to document it clearly.

salu133445 commented:
It does behave like a map on all the lists but does not return anything.

```python
>>> a, b, c = [0], [1], [2]
>>> map(lambda x: x.append(-1), [a,b,c])
<map at 0x23c92c3bf88>
>>> print(a, b, c)  # has no effect as map returns an iterator
[0] [1] [2]
>>> list(map(lambda x: x.append(-1), [a,b,c]))
[None, None, None]
>>> print(a, b, c)
[0, -1] [1, -1] [2, -1]
```

salu133445 commented:
In fact, `adjust_time` is not really pythonic either. It would be better if we could iterate over all timestamped objects.

salu133445 commented:
I do think implementing `remove_duplicate`, `remove_invalid` and `sort` in this way is fine, and it seems to fix #49. For `map`, `filter` and `each` (and possibly `apply`), we could open another pull request and see if we can come up with a more consistent, pythonic way to implement them.

cifkao commented:
> In fact, `adjust_time` is not really pythonic either. It would be better if we could iterate over all timestamped objects.

If it weren't for the intricacies of adjusting durations...

cifkao commented:
OK, I removed the functions.

I replaced `_apply_list_op` with `_traverse_lists`, which seems more pythonic. This also allowed me to implement `flat`.

salu133445 commented on muspy/base.py:
```diff
@@ -652,43 +662,28 @@ def remove_invalid(
-        if recursive and is_complexbase:
-            for item in value:
-                item.remove_duplicate(recursive=recursive)
+            if isclass(attr_type) and issubclass(attr_type, Base):
+                value[:] = [item for item in value  # type: ignore
+                            if (isinstance(item, attr_type)
+                                and cast(Base, item).is_valid())]
+            else:
+                value[:] = [item for item in value  # type: ignore
+                            if isinstance(item, attr_type)]
```
Type checking can be simplified if we change the return signature into `Iterator[Tuple[str, List]]`.

cifkao replied:
I think there must be a bug in mypy, it was complaining about things whose type is actually known...

salu133445 reviewed (changes requested):
- The implementation of `_traverse_lists` looks good to me.
- I am not a fan of `flat` in music21. We should use a more obvious and pythonic name. For now, I can only come up with `lists()` and `lists_with_names()`. There should be better options.
- I already fixed #49 by 8899576, but I am not sure which would be faster

cifkao commented on muspy/base.py:
```diff
@@ -753,11 +735,23 @@ def sort(
+
         return self
+
+    @property
+    def flat(self) -> Iterable:
+        """A flat representation of this object. Iterating over it
+        yields all items in all list attributes inside this object
+        (recursively). Non-list attributes are not included.
+        """
+        return self._flat
+
+    def _flat_generator(self) -> Iterable:
+        value: list
+        for _, _, value in self._traverse_lists(attr=None, recursive=True):
+            yield from value
```
Right, I didn't realize that.

How about this? `flat` is now reusable, and it's always the same object, so it's not a problem to have it as a property.

If you still insist on it being a function, why not call it e.g. `walk_lists`?

salu133445 replied:
I just want it to behave more like `dict.items()` or `dict.values()` rather than `numpy.ndarray.flat` or `music21.stream.Stream.flat`. It's quite common in Python to have a function that returns an iterator.

cifkao replied:
True. Is `walk_lists()` OK then? The name should give a hint that the return value is good only for (one) iteration.

BTW `dict.items()` and `dict.values()` also return objects that support more than just iteration.

cifkao replied:
Anyway, let's not overthink this. `lists()` with an optional `recursive` parameter (maybe `False` by default?) is OK too.

cifkao replied:
Actually, `lists()` suggests that it would iterate over lists. But my `flat` iterates over list items, not lists themselves. So we need a better name.

cifkao replied:
I made it `list_items()` and added `list()` to iterate over the actual lists.

cifkao commented on muspy/base.py:
```diff
@@ -537,6 +545,35 @@ def __add__(
+                yield from item._traverse_lists(  # type: ignore
+                    attr=None, recursive=recursive)
+
+        yield (attr, attr_type, getattr(self, attr))
```
Yes, I would definitely keep this and internal function.

cifkao commented:
OK, I replaced `flat` with `list_items()` and also added `lists()`.

Example:
```python
>>> list(m.lists())
[('tempos', []),
 ('key_signatures', []),
 ('time_signatures', []),
 ('downbeats', []),
 ('lyrics', []),
 ('annotations', []),
 ('tracks',
  [Track(program=33, is_drum=False, name='BB Bass', notes=[Note(time=0, pitch=34, duration=12, velocity=59), Note(time=18, pitch=43, duration=17, velocity=74), Note(time=36, pitch=39, duration=5, velocity=62), ...])])]

>>> list(m.list_items())
[Track(program=33, is_drum=False, name='BB Bass', notes=[Note(time=0, pitch=34, duration=12, velocity=59), Note(time=18, pitch=43, duration=17, velocity=74), Note(time=36, pitch=39, duration=5, velocity=62), ...])]

>>> pprint(m.lists(recursive=True))
[('tempos', []),                                                                                                                                                                                            
 ('key_signatures', []),                                                                                                                                                                                    
 ('time_signatures', []),                                                                                                                                                                                   
 ('downbeats', []),                                                                                                                                                                                         
 ('lyrics', []),                                                                                                                                                                                            
 ('annotations', []),                                                                                                                                                                                       
 ('notes',                                                                                                                                                                                                  
  [Note(time=0, pitch=34, duration=12, velocity=59),                                                                                                                                                        
   Note(time=18, pitch=43, duration=17, velocity=74),                                                                                                                                                       
   Note(time=36, pitch=39, duration=5, velocity=62),                                                                                                                                                        
   ...]),
 ('chords', []),
 ('lyrics', []),
 ('annotations', []),
 ('tracks',
  [Track(program=33, is_drum=False, name='BB Bass', notes=[Note(time=0, pitch=34, duration=12, velocity=59), Note(time=18, pitch=43, duration=17, velocity=74), Note(time=36, pitch=39, duration=5, velocity=62), ...])])]

>>> pprint(m.list_items(recursive=True))
[Note(time=0, pitch=34, duration=12, velocity=59),                                                                                                                                                          
 Note(time=18, pitch=43, duration=17, velocity=74),                                                                                                                                                         
 Note(time=36, pitch=39, duration=5, velocity=62),
 ...,
 Track(program=33, is_drum=False, name='BB Bass', notes=[Note(time=0, pitch=34, duration=12, velocity=59), Note(time=18, pitch=43, duration=17, velocity=74), Note(time=36, pitch=39, duration=5, velocity=62), ...])]
```
